### ISSUE TITLE
YOLOv8 struggles with detecting small objects

### ISSUE THREAD

**Issue by keisan1231** [ultralytics/ultralytics#14330]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hello,

I am using YOLOv8 to detect tennis balls (classified as "sports ball") in images. While the model works well for normally sized objects, it struggles to detect the tennis balls when they appear small in the image. If I crop and enlarge the area containing the small ball, the model detects it, but the quality of the image degrades, and this workaround is not practical for my application. (Please see below.)

It seems like YOLOv8 has limitations in detecting very small objects. I would like to understand if this is a known limitation of YOLOv8 and if there are any recommended approaches to improve small object detection. Here are some specific questions:

- Resolution: Should I use higher resolution input images to improve the detection of small objects? If so, are there recommended practices for handling higher resolution images with YOLOv8?

- Anchor Boxes: Is it possible to adjust the anchor boxes to better detect small objects? If yes, how can I modify the anchor boxes to improve small object detection?

- Training Data: Would increasing the amount of training data with small objects improve detection? Are there specific data augmentation techniques recommended for this purpose?

- Model Configuration: Are there any specific configurations or parameters within YOLOv8 that can be tuned to enhance the detection of small objects?

- Other Recommendations: Are there any other strategies or techniques that can help improve the detection of small objects with YOLOv8?

Thank you for your help!

| normal | enlarged image|
| ---- | ---- |
| input <img width="400" src="https://github.com/ultralytics/ultralytics/assets/46130313/b60bbe5c-1bc6-44c9-9b0a-47dfe14a518c" > | input <img width="400" src="https://github.com/ultralytics/ultralytics/assets/46130313/257cd816-eef4-4095-b596-f0ce2e887558" > |
| result <img width="400" src="https://github.com/ultralytics/ultralytics/assets/46130313/07b9f2a3-4b56-4eb2-a36e-f9b5d4d4ff00" > | result <img width="400" src="https://github.com/ultralytics/ultralytics/assets/46130313/508136cb-a744-428d-a9ac-a9aede9a4e6d" > |

**Comment by glenn-jocher** [comment]
@keisan1231 hello,

Thank you for reaching out and providing detailed information about the issue you're encountering with detecting small objects using YOLOv8. Detecting small objects can indeed be challenging, but there are several strategies you can employ to improve performance:

1. **Resolution**:
   - **Higher Resolution**: Increasing the input image resolution can help the model detect smaller objects. YOLOv8 can handle higher resolution images, but be mindful of the trade-off with processing speed and memory usage. You can set the image size using the `imgsz` parameter during training and inference.
     ```python
     # Example for training with higher resolution
     model.train(data='data.yaml', imgsz=1280)
     ```

2. **Anchor Boxes**:
   - **Custom Anchor Boxes**: Adjusting the anchor boxes can improve detection of small objects. You can use the `autoanchor` feature to automatically calculate optimal anchor boxes for your dataset.
     ```python
     # Example for training with autoanchor
     model.train(data='data.yaml', autoanchor=True)
     ```

3. **Training Data**:
   - **Increase Training Data**: More training data with small objects can significantly improve detection performance. Ensure your dataset includes a variety of images with small objects.
   - **Data Augmentation**: Techniques like mosaic augmentation, random cropping, and scaling can help the model learn to detect small objects better.
     ```yaml
     # Example augmentation settings in your data.yaml
     augment:
       mosaic: 1.0
       scale: 0.5
       crop: 0.5
     ```

4. **Model Configuration**:
   - **Hyperparameters**: Tuning hyperparameters such as learning rate, batch size, and epochs can also help. You might need to experiment with these to find the optimal settings for your specific use case.
   - **Smaller Strides**: Using a model with smaller strides can help detect smaller objects. You can modify the model architecture to use smaller strides in the early layers.

5. **Other Recommendations**:
   - **Feature Pyramid Networks (FPN)**: Using models that incorporate FPN can help improve small object detection by combining low-level and high-level features.
   - **Post-Processing**: Applying techniques like Non-Maximum Suppression (NMS) with lower IoU thresholds can help retain more small object detections.

If you haven't already, please ensure you are using the latest version of the Ultralytics YOLOv8 package, as updates often include improvements and bug fixes that could be beneficial.

For further guidance, you can refer to our [Common Issues Guide](https://docs.ultralytics.com/guides/yolo-common-issues/) which provides additional troubleshooting tips and best practices.

Feel free to share a minimum reproducible example if the issue persists, as it will help us provide more targeted assistance. You can find more information on creating a reproducible example [here](https://docs.ultralytics.com/help/minimum_reproducible_example).

Best of luck with your project, and don't hesitate to reach out if you have any more questions! 😊

**Comment by keisan1231** [comment]
@glenn-jocher 

Thank you for your reply. Currently, I do not have enough training data, so I would like to avoid training the model myself. In that case, which of the existing YOLOv8 models (n, s, m, l, x) would be most helpful for detecting small tennis balls? I plan to eventually use it in iOS with CoreML.

Also, is it possible to use FPN (Feature Pyramid Networks) with the CoreML format?

**Comment by glenn-jocher** [comment]
Hello @keisan1231,

Thank you for your detailed questions!

1. **Choosing the Right YOLOv8 Model**:
   - For detecting small objects like tennis balls, the larger models (YOLOv8-l or YOLOv8-x) generally perform better due to their higher capacity and more detailed feature extraction. However, they also require more computational resources. If you plan to deploy on iOS, you might need to balance performance and resource usage.

2. **Using FPN with CoreML**:
   - YOLOv8 models inherently utilize advanced feature extraction techniques, including FPN-like structures, to improve detection across scales. When you export a YOLOv8 model to CoreML, these architectural benefits are retained. You can export your model to CoreML using the following command:
     ```bash
     yolo export model=yolov8x.pt format=coreml
     ```

3. **Pre-trained Models**:
   - If you prefer not to train the model yourself, you can leverage pre-trained YOLOv8 models available in the Ultralytics repository. These models are trained on large datasets and can be fine-tuned for specific tasks if needed.

Feel free to experiment with the different model sizes to find the best balance for your application. If you encounter any issues or need further assistance, don’t hesitate to reach out!

Best of luck with your project! 😊

**Comment by keisan1231** [comment]
@glenn-jocher 

## New Question
In YOLO, the model detects all objects captured by the camera. However, I would like to specialize in sports balls to save memory and make it easier to use on iOS. Is this possible?

By this, I don't simply mean not outputting anything other than the class name "sports ball" to the UI. I mean specializing the prediction model itself for sports balls without additional training.

## Background

Thank you for your advice. Using yolov8x with imgsz=1280, I was able to recognize sports balls even in images taken with a wide-angle camera from a distance.
My current concern is the memory usage when converting this to CoreML and implementing real-time detection with a camera.

```sh
yolo predict model=yolov8x.pt source='wide.jpg' save=True project='runs' imgsz=1280
```

| input image | output image|
| ---- | ---- |
| <img width="400" src="https://github.com/user-attachments/assets/2cdcf858-8e64-4ca4-a51c-017afcc3c465" > | <img width="400" src="https://github.com/user-attachments/assets/4b7133dd-8f7d-417b-8532-81ef0bae4074" > |

**Comment by glenn-jocher** [comment]
It's great to hear that YOLOv8x with higher resolution is working well for your use case. To specialize the model for detecting only sports balls without additional training, you can filter the detections to include only the "sports ball" class during inference. This approach will help reduce memory usage and streamline the model for your specific application. When exporting to CoreML, ensure you optimize the model for mobile deployment by using the `optimize` and `half` parameters. This will help manage memory usage and improve performance on iOS devices. If you encounter any issues, please check for reproducibility with the latest package versions.

**Comment by curtis18** [comment]
> * Hyperparameters

Hello @glenn-jocher, could you tell us how to change to strides appropriately?

**Comment by glenn-jocher** [comment]
@curtis18 the model strides are fixed at 8, 16, 32, but you can effectively reduce or increase the grid size of an image just by predicting at lower or higher resolution, i.e. `yolo predict imgsz=1280` will have 4x the output grid points of the default imgsz=640.

**Comment by keisan1231** [comment]
@glenn-jocher  Thank you for the advice. We have decided to train with custom data. We used a service called Roboflow for annotation. Is it possible to export this trained model as YOLOv8 CoreML?

**Comment by glenn-jocher** [comment]
Yes, you can export your custom-trained model to CoreML using the following command: `yolo export model=path/to/your_model.pt format=coreml`.

**Comment by curtis18** [comment]
> @curtis18 the model strides are fixed at 8, 16, 32, but you can effectively reduce or increase the grid size of an image just by predicting at lower or higher resolution, i.e. `yolo predict imgsz=1280` will have 4x the output grid points of the default imgsz=640.

Hi @glenn-jocher, I tried to use yolo predict imgsz=1280, the result of predicting small object is not well. Maybe the Yolov8 model training input size is 640, and the prediction image needs to be downsampling for prediction, am I correct? By the way, may I know why imgsz=1280 will have 4x the output grid points of the default imgsz=640, but not 2x? how to calculate it? Thank you so much.

**Comment by glenn-jocher** [comment]
Hi @curtis18, you're correct that the model's performance can be influenced by the training input size. The grid size scales with the image size, so doubling the image dimensions (from 640 to 1280) results in a 4x increase in grid points (since both width and height are doubled). This is because the grid points are calculated as (width/stride) * (height/stride).

**Comment by keisan1231** [comment]
@glenn-jocher Thanks a lot🙏

**Comment by YOOHYOJEONG** [comment]
@glenn-jocher 
Hi,   
I faced the same problem.
   
You need to learn and detect very small bboxes from HD resolution images.
   
I tried changing the learning image size to 1280, but the effect is minimal.
   
   
Anchor Boxes:
- Custom Anchor Boxes: Adjusting the anchor boxes can improve detection of small objects. You can use the autoanchor feature to automatically calculate optimal anchor boxes for your dataset.   
```python
# Example for training with autoanchor
model.train(data='data.yaml', autoanchor=True)
```

This doesn't seem to be valid at the moment. Are there any other parameters that work like this?

**Comment by glenn-jocher** [comment]
Hi there! For small object detection, we recommend using our anchor-free models like YOLOv8 or YOLOv11 which often perform better on small objects without anchor tuning. The `autoanchor` functionality is now handled automatically during training when needed. You might also consider increasing mosaic augmentation and using copy-paste small object augmentation in your dataset. For details see our [AutoAnchor docs](https://docs.ultralytics.com/guides/hyperparameter-tuning/#autoanchor). 😊
